### PR TITLE
[kissfft] Update to 131.1.0

### PR DESCRIPTION
**Pull request by myd7349**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [ ] The "supports" clause reflects platforms that may be fixed by this new version
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/kissfft/fix-find-libpng.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+     include(FindPkgConfig)
++    if(PKG_CONFIG_FOUND)
+     pkg_check_modules(libpng REQUIRED IMPORTED_TARGET libpng)
+     add_kissfft_executable(psdpng psdpng.c)
+     target_link_libraries(psdpng PRIVATE PkgConfig::libpng)
++    else()
++        find_package(PNG REQUIRED)
++        add_kissfft_executable(psdpng psdpng.c)
++        target_link_libraries(psdpng PRIVATE PNG::PNG)
++    endif()
```
Comment (dg0yt) [inline ports/kissfft/fix-find-libpng.patch:17]: This seems to convoluted. Eiter make pkgconfig work everywhere (no patch needed), or always use `find_package`. Port png supports both.

File: ports/kissfft/portfile.cmake [hunk 2]
```diff
@@ -2,27 +2,112 @@ vcpkg_from_github(
...
+foreach(feature IN LISTS datatype_features)
+    if(feature IN_LIST FEATURES)
+        list(APPEND conflict_features "${feature}")
+    endif()
+endforeach()
+
+list(LENGTH conflict_features conflict_features_length)
+if(conflict_features_length GREATER_EQUAL 2)
+    message(FATAL_ERROR "These datatype features are conflict: '${conflict_features}'")
+endif()
```
Comment (dg0yt) [inline ports/kissfft/portfile.cmake:34]: Features must be additive. The datatype features seem to implement incompatible alternatives. I don't think this will be accepted.
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-features-to-implement-alternatives

File: ports/kissfft/portfile.cmake [hunk 3]
```diff
@@ -2,27 +2,112 @@ vcpkg_from_github(
...
 
 vcpkg_cmake_configure(
     SOURCE_PATH ${SOURCE_PATH}
-    OPTIONS_DEBUG
-        -DKF_INSTALL_HEADERS=OFF
-    OPTIONS_RELEASE
-        -DKF_INSTALL_HEADERS=ON
+    OPTIONS
+        -DKISSFFT_PKGCONFIG=ON
```
Comment (dg0yt) [inline ports/kissfft/portfile.cmake:51]: If it is to use pkg-config, the portfile will need to provide this (from `vcpkg_find_acquire_program(PKGCONFIG)`).

File: ports/kissfft/portfile.cmake [hunk 4]
```diff
@@ -2,27 +2,112 @@ vcpkg_from_github(
...
 
 vcpkg_cmake_install()
+
 vcpkg_copy_pdbs()
 
-vcpkg_cmake_config_fixup(
-    PACKAGE_NAME unofficial-${PORT}
-    CONFIG_PATH lib/cmake/unofficial-${PORT}
-)
+vcpkg_cmake_config_fixup(CONFIG_PATH "lib/cmake/${PORT}")
```
Comment (dg0yt) [inline ports/kissfft/portfile.cmake:62]: Do not use `${PORT}` but spell it out. The location is controlled by the package, not by the port.
(Another contributor just fell over this for case sensitive file systems.)

File: ports/kissfft/vcpkg.json [hunk 5]
```diff
@@ -12,5 +13,38 @@
...
     }
-  ]
+  ],
+  "default-features": [
+    "float"
+  ],
+  "features": {
+    "alloca": {
+      "description": "Use alloca instead of malloc"
+    },
```
Comment (dg0yt) [inline ports/kissfft/vcpkg.json:23]: Is this another feature implementing an alternative?

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/kissfft/portfile.cmake [hunk 6]
```diff
@@ -2,27 +2,112 @@ vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO mborgerding/kissfft
     REF 8f47a67f595a6641c566087bf5277034be64f24d
```
Comment (LilyWangLL) [inline ports/kissfft/portfile.cmake:4]:
```suggestion
    REF "${VERSION}"
```

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

File: ports/kissfft/fix-find-libpng.patch (hunk 1 above)
Comment (myd7349) [reply ports/kissfft/fix-find-libpng.patch]: @dg0yt Thanks for your review. That sounds reasonable to me. Since `psdpng` currently only supports Linux/macOS, so I will use pkgconfig.

File: ports/kissfft/portfile.cmake (hunk 3 above)
Comment (myd7349) [reply ports/kissfft/portfile.cmake:27]: When `KISSFFT_PKGCONFIG=ON`, it will generate a pkgconfig file `kissfft-float.pc` with pure CMake script.

Comment (myd7349) [reply ports/kissfft/portfile.cmake:27]: When `KISSFFT_PKGCONFIG=ON`, it will generate a pkgconfig file `kissfft-float.pc` with pure CMake script.

File: ports/kissfft/vcpkg.json (hunk 5 above)
Comment (myd7349) [reply ports/kissfft/vcpkg.json]: I

File: ports/kissfft/portfile.cmake (hunk 6 above)
Comment (myd7349) [reply ports/kissfft/portfile.cmake]: Thank you.

File: ports/kissfft/fix-install-dirs.patch [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -105,9 +105,7 @@ endif()
+ # Add GNUInstallDirs for GNU infrastructure before target)include_directories
+ #
+ 
+-if(CMAKE_SYSTEM_NAME MATCHES "^(Linux|kFreeBSD|GNU)$" AND NOT CMAKE_CROSSCOMPILING)
+-    include(GNUInstallDirs)
+-endif()
++include(GNUInstallDirs)
```
Comment (dg0yt) [inline ports/kissfft/fix-install-dirs.patch:12]: Is this change needed? (I don't know.)
GNUInstallDirs provides nice defaults and standard variable names, but `vcpkg_configure_cmake` also sets some of them explicitly.

**Review by dg0yt — COMMENTED** [review]
Thanks @myd7349.

File: ports/kissfft/vcpkg.json [hunk 8]
```diff
@@ -12,5 +13,17 @@
...
       "host": true
     }
-  ]
+  ],
+  "features": {
+    "openmp": {
+      "description": "Build kissfft with OpenMP support"
+    },
+    "tools": {
+      "description": "Build kissfft with SIMD support",
```
[the remark this answers is not in the document]
Comment (myd7349) [reply ports/kissfft/vcpkg.json]: Thanks. Nice catch.

**Comment by LilyWangLL** [comment]
Thanks for your PR. Feature `kissfft[openmp]:x64-windows` build failed with the following error:
```
CMake Error at CMakeLists.txt:202 (message):
  Don't know how to enable OpenMP for this compiler
```
The related codes, does this feature support `!windows`?
```
if(KISSFFT_OPENMP)
    if(CMAKE_C_COMPILER_ID MATCHES "GNU|Clang|AppleClang")
        if (NOT MSVC)
            target_compile_options(kissfft PRIVATE -fopenmp)
            if(${CMAKE_VERSION} VERSION_LESS "3.13.0")
                target_link_libraries(kissfft PRIVATE "-fopenmp")
            else()
                target_link_options(kissfft PRIVATE -fopenmp)
            endif()
        else()
            target_compile_options(kissfft PRIVATE "/openmp")
            if(${CMAKE_VERSION} VERSION_LESS "3.13.0")
                target_link_libraries(kissfft PRIVATE "/openmp")
            else()
                target_link_options(kissfft PRIVATE "/openmp")
            endif()
        endif()
        set(KISSFFT_EXPORT_SUFFIX "-openmp")
        set(KISSFFT_OUTPUT_NAME "kissfft-${KISSFFT_DATATYPE}-openmp")
    else()
        message(FATAL_ERROR "Don't know how to enable OpenMP for this compiler")
    endif()
endif()
```

**Comment by LilyWangLL** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Comment by LilyWangLL** [comment]
And usage test failed on x64-windows.

```
    find_package(kissfft CONFIG REQUIRED)
    target_link_libraries(main PRIVATE kissfft::kissfft kissfft::kissfft-float)
```
The error message:
```
1> [CMake] CMake Error at CMakeProject18/CMakeLists.txt:13 (target_link_libraries):
1> [CMake]   Target "CMakeProject18" links to:
1> [CMake] 
1> [CMake]     kissfft::kissfft
1> [CMake] 
1> [CMake]   but the target was not found.
```
Please also fix it.

File: ports/kissfft/portfile.cmake [hunk 9]
```diff
@@ -21,22 +21,26 @@ if("tools" IN_LIST FEATURES)
...
+set(datatypes "float;double;int16_t;int32_t")
+foreach(datatype IN LISTS datatypes)
+    vcpkg_cmake_configure(
+        SOURCE_PATH ${SOURCE_PATH}
+        OPTIONS
+            -DKISSFFT_DATATYPE=${datatype}
+            -DKISSFFT_PKGCONFIG=ON
+            -DKISSFFT_TEST=OFF
+            -DKISSFFT_STATIC=${KISSFFT_STATIC}
+            ${FEATURE_OPTIONS}
```
Comment (dg0yt) [inline ports/kissfft/portfile.cmake:33]:
~~~
        LOGFILE_BASE "config-${TARGET_TRIPLET}-${datatype)"
~~~

File: ports/kissfft/portfile.cmake [hunk 10]
```diff
@@ -21,22 +21,26 @@ if("tools" IN_LIST FEATURES)
...
+    )
 
-vcpkg_cmake_install()
+    vcpkg_cmake_install()
```
Comment (dg0yt) [inline ports/kissfft/portfile.cmake:36]:
~~~
        LOGFILE_BASE "install-${TARGET_TRIPLET}-${datatype)"
~~~

File: ports/kissfft/portfile.cmake [hunk 11]
```diff
@@ -21,22 +21,26 @@ if("tools" IN_LIST FEATURES)
...
 
-vcpkg_cmake_configure(
-    SOURCE_PATH ${SOURCE_PATH}
-    OPTIONS
-        -DKISSFFT_PKGCONFIG=ON
-        -DKISSFFT_TEST=OFF
-        -DKISSFFT_STATIC=${KISSFFT_STATIC}
-        ${FEATURE_OPTIONS}
-)
+set(datatypes "float;double;int16_t;int32_t")
```
Comment (dg0yt) [inline ports/kissfft/portfile.cmake:24]: This is valid syntax, but I would prefer to see a list of string instead of a single string.
~~~cmake
 set(datatypes float double int16_t int32_t)
~~~

**Review by dg0yt — COMMENTED** [review]
Building them all is fine if they don't clash. There is already another port which is using this approach.

File: ports/kissfft/portfile.cmake (hunk 10 above)
Comment (myd7349) [reply ports/kissfft/portfile.cmake]: Thank you!

It seems that `vcpkg_cmake_install` doesn't recognize `LOGFILE_BASE`, so I use `vcpkg_cmake_build` instead.

File: ports/kissfft/usage [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    find_package(kissfft CONFIG REQUIRED COMPONENTS STATIC float)
+    find_package(kissfft CONFIG REQUIRED COMPONENTS STATIC double)
+    find_package(kissfft CONFIG REQUIRED COMPONENTS STATIC int16_t)
+    find_package(kissfft CONFIG REQUIRED COMPONENTS STATIC int32_t)
+
+    # When built as shared library:
+    find_package(kissfft CONFIG REQUIRED COMPONENTS SHARED float)
+    find_package(kissfft CONFIG REQUIRED COMPONENTS SHARED double)
+    find_package(kissfft CONFIG REQUIRED COMPONENTS SHARED int16_t)
+    find_package(kissfft CONFIG REQUIRED COMPONENTS SHARED int32_t)
```
Comment (dg0yt) [inline ports/kissfft/usage:13]: As a vcpkg user, I would like to avoid writing `STATIC` or `SHARED`. And it seems to be optional indeed:
https://github.com/mborgerding/kissfft/blob/8f47a67f595a6641c566087bf5277034be64f24d/kissfft-config.cmake.in#L3-L5

(We might even want to modify the config file to make the default value follow linkage at time of build, not time of use.)

**Review by dg0yt — CHANGES_REQUESTED** [review]
IIUC there are two usage scenarios:
~~~
kissfft provides CMake targets:

    # Using distinct targets
    find_package(kissfft CONFIG REQUIRED)
    target_link_libraries(main PRIVATE kissfft::kissfft-float)
    # Alternative targets: kissfft::kissfft-double, kissfft::int16_t, kissfft::int32_t

    # Using a distinct package component
    find_package(kissfft CONFIG REQUIRED COMPONENTS int32_t)
    target_link_libraries(main PRIVATE kissfft::kissfft)
    # Alternative components: float, double, int16_t

**Comment by LilyWangLL** [comment]
Usage test pass on x64-windows.

**Comment by LilyWangLL** [comment]
Feature `kissfft[tools]:x64-windows` build failed with the following error:
```
CMake Error at scripts/cmake/vcpkg_copy_tools.cmake:36 (message):
  Couldn't find tool "fastconv":

      "/home/user/kissfft/packages/kissfft_x64-linux/bin/fastconv" does not exist
Call Stack (most recent call first):
  ports/kissfft/portfile.cmake:52 (vcpkg_copy_tools)
  scripts/ports.cmake:147 (include)
```
The tools in `bin` folder:

image

**Comment by LilyWangLL** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Review by LilyWangLL — APPROVED** [review]

**Comment by LilyWangLL** [comment]
All features passed on `x64-linux`.
